src/mongoc/mc_topology_logics.erl

```diff
@@ -19,7 +19,7 @@
 -define(NOT_MAX(E, M), E =/= undefined, M =/= undefined, E < M).
 
 %% API
--export([update_topology_state/2, init_seeds/4, init_seeds/1]).
+-export([update_topology_state/2, init_seeds/4, init_seeds/1, validate_server_and_config/3, server_type/1]).
 
 update_topology_state(#mc_server{type = SType, pid = Pid}, State = #topology_state{type = sharded}) when ?NON_SHARDED(SType) -> %% SHARDED
   exit(Pid, kill),
```

Comment:
Should we print some logs here, or exit the process with a more meaningful signal like `topology_type_miss_match`?

Reply:
If I did everything right we shouldn't ever hit these exit calls within `update_topology_state` anymore because of the new checks I added which prevents the `mc_topology` gen_server from successfully completing an `init` call. These specific methods **should** be dead code now that are never reached.

Though perhaps that is wishful thinking? Perhaps something could change in the server config after we've tried to connect and suddenly this code might be reachable because IIRC there are repeated checks against the server/cluster which might make these code paths active again on the next check **if** some change has been made. If that does happen we're right back into the same looping problem this new code tries to solve.

I am not 100% confident that nothing could change out from 'underneath' us, or anyone using the client.

Either way changing the reason for the kill doesn't help us. This is the method which handles the message for the linked process:
```
handle_info({'DOWN', MRef, _, _, _}, State = #topology_state{topology_opts = Topts, worker_opts = Wopts, servers = Tab}) ->
  case ets:match(Tab, #mc_server{pid = '$1', host = '$2', mref = MRef, _ = '_'}) of
    [[Pid, Host]] ->
      true = ets:delete(Tab, Pid),
      mc_topology_logics:init_seeds([Host], Tab, Topts, Wopts),
      {noreply, State};
    [] ->
      {noreply, State}
  end;
```

No reason is checked and the cycle immediately begins again with the `init_seeds` call so it wouldn't actually buy us anything.

Logging though is something we could do.

Another possibility if we want to slow down the loop just in case it does happen again would be for the `mc_topology` gen_server to send itself a delayed message to tell it to init things again rather than doing it immediately.

If we do both of those things we'd get a log and a much slower loop.

Reply:
Sorry for the delay.

I think we'd better make the loop slower, i.e. every 1 second. And we need to give a more human-readable log message if the gen_server crashes because the config has been changed in the server side.